Cross-platform suite-local fixtures

armanbilge opened the pull request:
This PR introduces a `ResourceSuiteLocalDeferredFixture[A]`. It has identical semantics to JVM-only `ResourceSuiteLocalFixture[A]` (i.e., allocate once per suite) but instead of providing `A` provides an `IO[A]` that is semantically blocked until the resource is allocated. This allows the fixture to be used on both JVM and JS.

RaasAhsan commented on common/shared/src/test/scala/munit/CatsEffectFixturesSpec.scala:
```diff
@@ -0,0 +1,61 @@
+
+  override def afterAll(): Unit = {
+    assertEquals(acquired, 1)
+    // assertEquals(released, 1) // Release is async, no way to check?
```
Yeah this is really my only question here. since there's no way to enforce backpressure between the release of the suite-local fixture and the acquisition of the next, you could run into some weird interactions between different suites. for example, if your fixture is a network server that binds to the same port. suite A might not release its port before the suite B attempts to acquire

I think this is still really useful for javascript to have, so this may just need to be a tradeoff that we warn users about

armanbilge replied:
> I think this is still really useful for javascript to have, so this may just need to be a tradeoff that we warn users about

I'm mindful of this issue too, but exactly as you said. For plenty of non-interacting resources should be fine.

armanbilge replied:
It's particularly evil, but this problem could be solved via some sort of global semaphore right?

RaasAhsan replied:
That sounds like it should work, but it would inhibit parallel suite execution (only the suites which use this feature) for users who opt in. which again may be fine and just something we warn users about

the fixtures do have names, so ideally you could split the global semaphore by name, but that relies on the assumption that interacting resources have the same name

armanbilge replied:
Right, but on JS we don't have parallel execution anyway. So maybe the trick is to use this implementation for JS, your blocking implementation for JVM, and just provide them under the same interface?

armanbilge replied:
Btw, how is it possible to run suites in parallel on JVM if they are using fixtures, couldn't you run into the same interaction problems? Does it actually run in parallel?

milanvdm replied:
Why wouldnt we able to use `unsafeRunSync` instead of `unsafeRunAndForget`? Wouldn't this make sure the `Resource` will be released after the `afterAll` and prevent the parallel execution problems?

armanbilge replied:
Yes you are right! But there is no unsafeRunSync for JS because it's single threaded ... blocking your only thread would permanently deadlock you :)

armanbilge replied:
Also apologies, I didn't realize you are a maintainer here as well, many thanks for reviewing!! Much appreciated.

milanvdm replied:
Would it make sense to then split this Suite into a jvm and js version? As mentioned, it might cause quite some problems when running tests in parallel if the resource isnt closed :)

armanbilge replied:
Yeah, I'm not sure. There's already a JVM-only `ResourceSuiteLocalFixture[A]` which has exactly the correct semantics. The goal here was to get something that's supported on both JVM and JS. We _could_ make the JVM version of `ResourceSuiteLocalDeferredFixture[A]` safer, but then its semantics wouldn't match the JS version ... thoughts?

armanbilge replied:
Maybe we should call it `UnsafeResourceSuiteLocalFixture` to emphasize these hazards.

milanvdm replied:
Im still a bit confused :)

We have a working `ResourceSuiteLocalFixture` for the JVM only. As far as I understand, you want to add similar functionality but for JS?

I don't really know anything about ScalaJS and how to create libraries for it, but I don't get why we need to create something that works on both the JVM and JS? 
Because, this shared logic on the JVM, doesn't really work as one would expect. And since we already have something that works the JVM, why add something shared?

So in my mind, that has no JS knowledge, can this logic simply be added to the JS part of this library? Or is there some value in having something shared?

armanbilge replied:
Good questions!

> Or is there some value in having something shared

Yes, absolutely. If it is not shared, then if you want to test your code on both JVM and JS you'd have to write your tests twice, separately for each platform. For a concrete example of this consider the http4s test suite which now needs to support both JVM and JS platforms where in fact I've already been using this construct:
https://github.com/http4s/http4s/blob/main/testing/shared/src/test/scala/org/http4s/Http4sSuite.scala#L68

milanvdm replied:
Then I would indeed go for the `UnsafeResourceSuiteLocalFixture` :)

Does the `CatsEffectFunFixtures` not give you the functionality you desire?

armanbilge replied:
`CatsEffectFunFixtures` does work on JVM and JS, but it allocates/tears down the resource once per test. The `SuiteLocal` fixtures allocate/tear down once per suite.

armanbilge commented:
@danicheg would you have a moment to review this please? I'd find this incredibly useful.

danicheg commented on ce2/shared/src/main/scala/munit/CatsEffectSuite.scala:
```diff
@@ -47,4 +47,11 @@ abstract class CatsEffectSuite
       { case e: SyncIO[_] => Future(e.unsafeRunSync())(munitExecutionContext) }
     )
 
+  private[munit] def unsafeRunAndForget[A](ioa: IO[A]): Unit = ioa.unsafeRunAsyncAndForget()
```
I have concerns about using `unsafeRunAndForget`. That would probably lead to some problems with releasing resources. Is it possible to use `unsafeToFuture` instead?

armanbilge replied:
You are right, using unsafeRunAnd forget is not ideal. But munit doesn't support using Futures instead of fixtures AFAICT.

danicheg commented on ce2/shared/src/main/scala/munit/CatsEffectSuite.scala:
```diff
@@ -47,4 +47,11 @@ abstract class CatsEffectSuite
+}
+
+object CatsEffectSuite {
+  type Deferred[F[_], A] = cats.effect.concurrent.Deferred[F, A]
```
I vote to make this `private`

armanbilge commented:
@milanvdm @danicheg thank you both for your thoughtful reviews, I've pushed some changes which hopefully address both of your concerns.
1. Renamed `ResourceSuiteLocalDeferredFixture` to `UnsafeResourceSuiteLocalDeferredFixture`. Furthermore, added a detailed comment explaining its semantics and why it is unsafe.
2. Platformed the implementation for JVM and JS, so that on JVM it uses `unsafeRunSync()` and only uses `unsafeRunAndForget()` for JS. TBH I'm not convinced this really buys us anything since the caveats still apply to JS platform, but at least it may make things a bit better on JVM.

Let me know what you think!

danicheg reviewed (commented):
You did a huge job @armanbilge 👍🏻 
Sorry, but I was out from your discussion with @milanvdm at some point. 
But as I get the root problem is that `ResourceSuiteLocalFixture` currently not working on JS correctly. And adding the `ResourceSuiteLocalDeferredFixture` would mean that `ResourceSuiteLocalFixture` will still not working on JS.
Maybe it's possible to change `ResourceSuiteLocalFixture` to have a working version both on JVM and JS? For example, we could put implement details of `munit.CatsEffectFixtures.ResourceSuiteLocalFixture#apply` for JVM and JS separately. 
So we would have one suite to work with resource fixtures both on JVM and JS. 
What do you folks think?

armanbilge commented:
@danicheg Unfortunately, it is not possible to `ResourceSuiteLocalFixture` fixture on JS, even if we implement it separately. The reason is because a `Resource[IO, T]` allocates a `T` _asynchronously_ but the fixture setup must run _synchronounously_. The reason this works on JVM is because you can use `unsafeRunSync()` to wait for `T`: https://github.com/typelevel/munit-cats-effect/blob/main/common/jvm/src/main/scala/munit/CatsEffectFixtures.scala#L41

On JS, there is only one thread and therefore no `unsafeRunSync()` available. So, it is impossible to allocate a `T` synchronously. So, the trick I used here, is instead of providing a `T`, to provide an `IO[T]` that is backed by a `Deferred[IO, T]`. This is possible, because although you cannot allocate `T` synchronously, you can allocate `Deferred[IO, T]` synchronously and then provide an `IO[T]` that will semantically block until the deferred is set with the allocated `T`.

Notice that this requires making a source-incompatible change to the signature. Which is obviously not an _impossible_ thing to do, but it would be painful for any downstream currently using `ResourceSuiteLocalFixture`.

If we really want to fix all this craziness: would require a PR to `munit` itself to add support for asynchronous fixtures. That would solve all the problems discussed here. See https://github.com/scalameta/munit/issues/186. If you are interested in solving that issue, that would be really amazing!! Especially since so many Typelevel projects are using _both_ CE and munit.

danicheg commented:
@armanbilge thanks for the detailed answer. What about using `cats.effect.IO#unsafeRunAsync` instead of `unsafeRunSync` for JS? I've sketched the implementation of `ResourceSuiteLocalFixture` for JS: https://gist.github.com/danicheg/be9ed79cd6f5764899e7cc16fe23e48c
That seems semantic is the same as for `ResourceSuiteLocalDeferredFixture`. Or maybe I missed something.

armanbilge commented:
Cool, thanks for giving it a try!! Have you tried running it against the test suite? 😉 

My concern is that there is no guarantee when [`unsafeRunAsync`](https://gist.github.com/danicheg/be9ed79cd6f5764899e7cc16fe23e48c#file-resourcesuitelocalfixtureplatform-scala-L24) will complete. So, a test may start while `value` is still `None`.

armanbilge commented:
Also, you might find the implementation of `unsafeRunAndForget` interesting :)
https://github.com/typelevel/cats-effect/blob/series/3.x/core/shared/src/main/scala/cats/effect/IO.scala#L709

danicheg commented:
@armanbilge yep, the tests were passed ._. But I'm not sure if this is legal though.

armanbilge commented:
Sounds like the test suite is not rigorous enough! 😬 We should fix that. For example, try using your fixture to allocate this resource: `Resource.eval(IO.pure("hello").delayBy(1.second))`.

danicheg commented:
@armanbilge sorry, I've double-checked it and tests didn't run at all. To be honest, they are not running and for the current version of `ResourceSuiteLocalFixture` for JS 🤷🏻

armanbilge commented:
That's probably because `ResourceSuiteLocalFixture` is in the JVM-only source directory?

danicheg commented:
@armanbilge yep, it is. So, version with `unsafeRunAsync` fails on the same case of `afterAll` as version on `Deffered` - `assertEquals(released, 1)`. But adding delay has lead to instantiation problems. So, you are definitely right.

armanbilge commented:
@danicheg thanks for looking into that. So, it definitely sounds like the current test suite is not sufficient then. I'll push a fix for that.

armanbilge commented:
@danicheg I added an `IO.sleep` to the fixture tests so that they properly test for asynchronous fixture allocation. Should be good to go!
